src/main/java/com/limechain/runtime/WasmSectionUtils.java

```diff
@@ -1,124 +1,120 @@
 
 @UtilityClass
 public class WasmSectionUtils {
-    private static final byte[] RUNTIME_VERSION_SECTION_NAME = "runtime_version".getBytes();
-    private static final byte[] RUNTIME_APIS_SECTION_NAME = "runtime_apis".getBytes();
+    private static final String RUNTIME_VERSION_SECTION_NAME = "runtime_version";
+    private static final String RUNTIME_APIS_SECTION_NAME = "runtime_apis";
+    private static final String MEMORY_IMPORT_NAME = "memory";
+    private static final String ENV_MODULE_NAME = "env";
+
+    private static final int DEFAULT_MEMORY_PAGES = 2048;
 
     /**
      * Parses the runtime version if both wasm custom sections ("runtime_apis" and "runtime_version") are present.
+     *
      * @param wasmBinary the wasm blob
      * @return the parsed RuntimeVersion if both custom sections are present, null otherwise
      */
     @Nullable
-    public RuntimeVersion parseRuntimeVersionFromCustomSections(byte[] wasmBinary) {
-        // byte value of \0asm concatenated with 0x1, 0x0, 0x0, 0x0
-        // taken from Smoldot: https://github.com/smol-dot/smoldot/blob/21be5a1abaebeaf7270a744485b4551da8636fb1/lib/src/executor/host/runtime_version.rs#L97
-        byte[] searchKey = new byte[]{0x00, 0x61, 0x73, 0x6D, 0x1, 0x0, 0x0, 0x0};
-
-        int searchedKeyIndex = ByteArrayUtils.indexOf(wasmBinary, searchKey);
-        if (searchedKeyIndex < 0) {
-            throw new WasmRuntimeException("Key not found in runtime code.");
-        }
+    public RuntimeVersion parseRuntimeVersionFromBinary(byte[] wasmBinary) {
+        Module moduleWithCustomSections = toModuleWithSections(wasmBinary, SectionId.CUSTOM);
 
-        WasmCustomSection runtimeApis = findRuntimeApisCustomSection(wasmBinary);
-        WasmCustomSection runtimeVersion = findRuntimeVersionCustomSection(wasmBinary);
+        UnknownCustomSection runtimeApis = (UnknownCustomSection) moduleWithCustomSections
+            .customSection(RUNTIME_APIS_SECTION_NAME);
+        UnknownCustomSection runtimeVersion = (UnknownCustomSection) moduleWithCustomSections
+            .customSection(RUNTIME_VERSION_SECTION_NAME);
 
         // If we're missing any of the two custom sections, fallback to `Core_runtime`
         if (Objects.isNull(runtimeApis) || Objects.isNull(runtimeVersion)) {
             return null;
         }
 
         // Read as many api versions as there are (we don't know the length for some reason).
-        ApiVersions apiVersions = ApiVersions.decodeNoLength(runtimeApis.content());
+        ApiVersions apiVersions = ApiVersions.decodeNoLength(runtimeApis.bytes());
 
         // Construct the runtime version partially uninitialized,
         // and then immediately set its apis we've read from the custom section.
-        RuntimeVersion version = new RuntimeVersionReader().read(new ScaleCodecReader(runtimeVersion.content()));
+        RuntimeVersion version = new RuntimeVersionReader().read(new ScaleCodecReader(runtimeVersion.bytes()));
```

Comment:
Nit to my old self :smile: :
```suggestion
RuntimeVersion version = ScaleUtils.Decode.decode(runtimeVersion.bytes(), new RuntimeVersionReader());
```